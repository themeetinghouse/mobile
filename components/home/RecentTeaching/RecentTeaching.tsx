import React, { useState } from 'react';
import { Text, View } from 'native-base';
import { Image, StyleSheet, Dimensions, TouchableWithoutFeedback } from 'react-native';
import { Theme, Style } from '../../../Theme.style';
import IconButton from '../../buttons/IconButton';
import moment from 'moment';
import { useNavigation } from '@react-navigation/native';
import { GetNotesQuery } from '../../../services/API';
import { MainStackParamList } from '../../../navigation/AppNavigator';
import { StackNavigationProp } from '@react-navigation/stack';
import ActivityIndicator from '../../../components/ActivityIndicator';
import { TouchableHighlight } from 'react-native-gesture-handler';
import { VideoData } from '../../../types';
import FallbackImage from '../../../components/FallbackImage';

const screenWidth = Dimensions.get('window').width;

const style = StyleSheet.create({
    container: {
        alignItems: "center",
        backgroundColor: Theme.colors.black,
    },
    teachingImage: {
        width: screenWidth,
        height: (9 / 16) * screenWidth,
        position: 'absolute'
    },
    seriesImage: {
        alignSelf: 'center',
        width: screenWidth * 0.2133,
        height: screenWidth * 0.256,
        marginBottom: 24
    },
    seriesImageWithTeachingImage: {
        marginTop: 32 + (0.75 * (9 / 16) * screenWidth),
    },
    episodeNumber: {
        color: Theme.colors.white,
        fontSize: Theme.fonts.smallMedium,
        fontFamily: Theme.fonts.fontFamilyBold,
        marginBottom: 4
    },
    title: {
        color: Theme.colors.white,
        fontSize: Theme.fonts.extraLarge,
        fontFamily: Theme.fonts.fontFamilyBold,
        marginBottom: 8,
        textAlign: "center",
    },
    subtitle: {
        color: Theme.colors.gray5,
        fontSize: Theme.fonts.small,
        marginBottom: 14
    },
    description: {
        ...Style.cardDescription,
        ...{
            color: Theme.colors.gray5,
            fontSize: Theme.fonts.medium,
            marginBottom: 32,
            textAlign: "center",
        }
    }
})

interface Props {
    note: GetNotesQuery['getNotes'];
    teaching: VideoData;
}

export default function RecentTeaching({ note, teaching }: Props): JSX.Element {
    const navigation = useNavigation<StackNavigationProp<MainStackParamList>>();
    const [fullDescription, setFullDescription] = useState(false);

    if (teaching && moment(teaching.publishedDate).isSameOrAfter(moment(note?.id))) {
        const series = teaching.series || { title: "" };
        const seriesImageUri = series.title
            ? `https://themeetinghouse.com/static/photos/series/adult-sunday-${teaching?.series?.title?.replace("?", "")}.jpg`
            : "https://www.themeetinghouse.com/static/NoCompassionLogo.png";

        const teachingImage = teaching?.Youtube?.snippet?.thumbnails?.standard ?? null;

        const openNotes = () => {
            navigation.push("NotesScreen", { date: moment(teaching.publishedDate as string).format("YYYY-MM-DD") })
        }

        return (
            <View style={style.container} testID='teaching-video'>
                {teachingImage?.url ?
                    <View style={{ width: '100%' }}>
                        <TouchableWithoutFeedback onPress={() => {
                            navigation.navigate("Main", { screen: "Teaching" });
                            navigation.navigate("SermonLandingScreen", { item: teaching })
                        }} >
                            <Image style={style.teachingImage} source={{ uri: teachingImage.url }} />
                        </TouchableWithoutFeedback>
                        <Image style={[style.seriesImage, style.seriesImageWithTeachingImage]} source={{ uri: seriesImageUri }} />
                    </View>
                    : <Image style={style.seriesImage} source={{ uri: seriesImageUri }}></Image>
                }
                <View style={{ marginHorizontal: 16, alignItems: 'center' }} >
                    <Text style={style.title}>{teaching.episodeTitle}</Text>
                    <View style={{ flexDirection: "row" }}>
                        <Text style={style.subtitle}>{moment(teaching.publishedDate as string).format("MMMM D, YYYY")} {teaching?.speakers?.items?.[0] ? 'by ' : ""}</Text>
                        <TouchableHighlight style={style.subtitle} onPress={() => navigation.push("Main", { screen: "More", params: { screen: "TeacherProfile", params: { staff: { idFromTeaching: teaching?.speakers?.items?.[0]?.speaker?.id } } } })}>
                            <Text style={[style.subtitle, { textDecorationLine: "underline" }]}>{teaching?.speakers?.items?.[0] ? `${teaching?.speakers?.items?.[0]?.speaker?.id}` : ""}</Text>
                        </TouchableHighlight>
                    </View>
                    <Text style={style.description} numberOfLines={fullDescription ? undefined : 2} ellipsizeMode='tail' onPress={() => setFullDescription(!fullDescription)}>{teaching.description}</Text>
                </View>
                {moment(teaching.publishedDate).isSame(moment(note?.id)) && <View testID='notes-button'>
                    <IconButton icon={Theme.icons.white.notes} label="Notes" onPress={openNotes} />
                </View>}
            </View>
        )
    }

    if (note) {
        const seriesImageUri = `https://themeetinghouse.com/static/photos/series/adult-sunday-${note?.seriesId.replace("?", "")}.jpg`
        const openNotes = () => {
            navigation.navigate("NotesScreen", { date: note.id })
        }
        return (
            <View style={style.container} testID='teaching-notes'>
                <FallbackImage style={style.seriesImage} uri={seriesImageUri} catchUri="https://www.themeetinghouse.com/static/NoCompassionLogo.png" />
                <View style={{ marginHorizontal: 16, alignItems: 'center' }} >
                    <Text style={style.title}>{note.title}</Text>
                    <Text style={style.subtitle}>{moment(note.id).format("MMMM D, YYYY")}</Text>
                    <Text style={style.description} numberOfLines={fullDescription ? undefined : 2} ellipsizeMode='tail' onPress={() => setFullDescription(!fullDescription)} >{note.episodeDescription}</Text>
                </View>
                <View testID='notes-button'>
                    <IconButton icon={Theme.icons.white.notes} label="Notes" onPress={openNotes} />
                </View>
            </View>
        )
    }

    return <ActivityIndicator />
}
