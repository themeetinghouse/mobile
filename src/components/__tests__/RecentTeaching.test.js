import React from 'react';
import { render } from '@testing-library/react-native';
import RecentTeaching from '../home/RecentTeaching';

export const mockTeaching = [
  {
    Youtube: {
      snippet: {
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/QzSrQpC8A6A/default.jpg',
          },
          high: {
            url: 'https://i.ytimg.com/vi/QzSrQpC8A6A/hqdefault.jpg',
          },
          maxres: {
            url: 'https://i.ytimg.com/vi/QzSrQpC8A6A/maxresdefault.jpg',
          },
          medium: {
            url: 'https://i.ytimg.com/vi/QzSrQpC8A6A/mqdefault.jpg',
          },
          standard: {
            url: 'https://i.ytimg.com/vi/QzSrQpC8A6A/sddefault.jpg',
          },
        },
      },
    },
    YoutubeIdent: 'QzSrQpC8A6A',
    audioURL:
      'https://media.themeetinghouse.com/podcast/audio/2020/2020-08-30-1246-sermon.mp3',
    description:
      'In John 8, Jesus confronts the systems we can rely on for security. Instead Jesus invites us into real security in him and in community - one of the gospel promises of a new way of relating to God and each other.',
    episodeNumber: 38,
    episodeTitle: 'Real Security',
    id: 'QzSrQpC8A6A',
    length: '29',
    notesURL: 'https://www.themeetinghouse.com/notes/2020-08-30',
    publishedDate: '2020-08-30',
    series: {
      id: 'Jesus by John',
      title: 'Jesus by John',
    },
    seriesTitle: 'Jesus by John',
    speakers: {
      items: [
        {
          speaker: {
            id: 'Lisa Goetze',
          },
        },
      ],
    },
    videoTypes: 'adult-sunday',
    videoURL:
      'https://media.themeetinghouse.com/vpodcast/2020/2020-08-30-1246-video.mp4',
    viewCount: '393',
  },
  {
    Youtube: {
      snippet: {
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/V5An4R0F1BY/default.jpg',
          },
          high: {
            url: 'https://i.ytimg.com/vi/V5An4R0F1BY/hqdefault.jpg',
          },
          maxres: {
            url: 'https://i.ytimg.com/vi/V5An4R0F1BY/maxresdefault.jpg',
          },
          medium: {
            url: 'https://i.ytimg.com/vi/V5An4R0F1BY/mqdefault.jpg',
          },
          standard: {
            url: 'https://i.ytimg.com/vi/V5An4R0F1BY/sddefault.jpg',
          },
        },
      },
    },
    YoutubeIdent: 'V5An4R0F1BY',
    audioURL:
      'https://media.themeetinghouse.com/podcast/audio/2020/2020-12-13-1261-sermon.mp3',
    description:
      'Tapping into joy as a way of living generously helps us and allows us to extend that to others, especially when Christmas can be a hard season.',
    episodeNumber: 2,
    episodeTitle: 'Comedy & A Posture of Generosity',
    id: 'V5An4R0F1BY',
    length: '42',
    notesURL: 'https://www.themeetinghouse.com/notes/2020-12-13',
    publishedDate: '2020-12-13',
    series: {
      id: 'Christmas in 3 Acts',
      title: 'Christmas in 3 Acts',
    },
    seriesTitle: 'Christmas in 3 Acts',
    speakers: {
      items: [
        {
          speaker: {
            id: 'Danielle Strickland',
          },
        },
      ],
    },
    videoTypes: 'adult-sunday',
    videoURL:
      'https://media.themeetinghouse.com/vpodcast/2020/2020-12-13-1261-video.mp4',
    viewCount: '511',
  },
  {
    Youtube: {
      snippet: {
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/default.jpg',
          },
          high: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/hqdefault.jpg',
          },
          maxres: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/maxresdefault.jpg',
          },
          medium: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/mqdefault.jpg',
          },
          standard: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/sddefault.jpg',
          },
        },
      },
    },
    YoutubeIdent: 'V_uARB3hhj0',
    audioURL:
      'https://media.themeetinghouse.com/podcast/audio/2020/2020-12-24-1263-sermon.mp3',
    description:
      'Through carols, art, and stories, we want to explore how this radical idea of God coming to us as a helpless baby changed everything, and how it continues to bring light, hope, and peace when this season of life feels dark.',
    episodeNumber: 1,
    episodeTitle: 'Christmas Eve at Home',
    id: 'V_uARB3hhj0',
    length: '54',
    notesURL: null,
    publishedDate: '2020-12-24',
    series: {
      id: 'Christmas Eve at Home',
      title: 'Christmas Eve at Home',
    },
    seriesTitle: 'Christmas Eve at Home',
    speakers: {
      items: [
        {
          speaker: {
            id: 'Bruxy Cavey',
          },
        },
      ],
    },
    videoTypes: 'adult-sunday',
    videoURL:
      'https://media.themeetinghouse.com/vpodcast/2020/2020-12-24-1263-video.mp4',
    viewCount: '2627',
  },
  {
    Youtube: {
      snippet: {
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/default.jpg',
          },
          high: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/hqdefault.jpg',
          },
          maxres: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/maxresdefault.jpg',
          },
          medium: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/mqdefault.jpg',
          },
          standard: {
            url: 'https://i.ytimg.com/vi/V_uARB3hhj0/sddefault.jpg',
          },
        },
      },
    },
    YoutubeIdent: 'V_uARB3hhj0',
    audioURL:
      'https://media.themeetinghouse.com/podcast/audio/2020/2020-12-24-1263-sermon.mp3',
    description:
      'Through carols, art, and stories, we want to explore how this radical idea of God coming to us as a helpless baby changed everything, and how it continues to bring light, hope, and peace when this season of life feels dark.',
    episodeNumber: 1,
    episodeTitle: 'Christmas Eve at Home',
    id: 'V_uARB3hhj0',
    length: '54',
    notesURL: null,
    publishedDate: '2020-12-24',
    series: {
      id: 'Christmas Eve at Home',
      title: 'Christmas Eve at Home',
    },
    seriesTitle: 'Christmas Eve at Home',
    speakers: {
      items: [
        {
          speaker: {
            id: 'Bruxy Cavey',
          },
        },
      ],
    },
    videoTypes: 'adult-sunday',
    videoURL:
      'https://media.themeetinghouse.com/vpodcast/2020/2020-12-24-1263-video.mp4',
    viewCount: '2627',
  },
];

export const mockNotes = [
  {
    createdAt: '2020-09-06T05:38:19.616Z',
    episodeDescription:
      'A message to all Pastors, Overseers, Elders, Leaders, and Disciples of Jesus about how to lead and how to follow.',
    episodeNumber: 39,
    id: '2020-09-06',
    pdf:
      'http://media.themeetinghouse.com/podcast/handouts/2020-09-06-1247-notes.pdf',
    seriesId: 'Jesus by John',
    tags: [],
    title: 'Gateway Leadership',
    topics: null,
    updatedAt: '2020-09-13T01:32:57.295Z',
  },
  {
    createdAt: '2020-12-13T00:47:50.867Z',
    episodeDescription:
      'Tapping into joy as a way of living generously helps us and allows us to extend that to others, especially when Christmas can be a hard season.',
    episodeNumber: 2,
    id: '2020-12-13',
    pdf:
      'http://media.themeetinghouse.com/podcast/handouts/2020-12-13-1261-notes.pdf',
    seriesId: 'Christmas in 3 Acts',
    tags: [],
    title: 'Comedy & A Posture of Generosity',
    topics: null,
    updatedAt: '2020-12-13T01:24:50.455Z',
  },
  {
    createdAt: '2020-12-26T17:02:13.357Z',
    episodeDescription:
      "As Jesus people, we aren't the ones with all the answers - we're the ones with hope. We partner, wait, labour, and celebrate with the One who doesn't grow tired or weary as He ushers in his kingdom.",
    episodeNumber: 1,
    id: '2020-12-27',
    pdf:
      'http://media.themeetinghouse.com/podcast/handouts/2020-12-27-1263-notes.pdf',
    seriesId: '000',
    tags: [],
    title: 'Tired Feet. Hopeful Hearts.',
    topics: null,
    updatedAt: '2020-12-27T14:19:01.046Z',
  },
  {
    createdAt: '2020-12-19T20:19:27.356Z',
    episodeDescription:
      'The Christmas story as fairytale, full of imagery and dramatic action, draws us into the larger narrative of who God is, into a battle being won in beautiful and unexpected ways. ',
    episodeNumber: 3,
    id: '2020-12-20',
    pdf:
      'http://media.themeetinghouse.com/podcast/handouts/2020-12-20-1262-notes.pdf',
    seriesId: 'Christmas in 3 Acts',
    tags: [],
    title: 'Fairytale & A Posture of Mission',
    topics: null,
    updatedAt: '2020-12-19T20:53:21.692Z',
  },
];

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({
      navigate: jest.fn(),
    }),
  };
});

// Regular Sunday
test('On Sunday Sept 6, expect notes from Sunday Sept 6', () => {
  const { queryByTestId } = render(
    <RecentTeaching teaching={mockTeaching[0]} note={mockNotes[0]} />
  );

  expect(queryByTestId('teaching-notes')).toBeTruthy();
  expect(queryByTestId('teaching-video')).toBeFalsy();
  expect(queryByTestId('notes-button')).toBeTruthy();
});

// During middle of week
test('On Wednesday Dec 16, expect video from Sunday Dec 13', () => {
  const { queryByTestId } = render(
    <RecentTeaching teaching={mockTeaching[1]} note={mockNotes[1]} />
  );

  expect(queryByTestId('teaching-notes')).toBeFalsy();
  expect(queryByTestId('teaching-video')).toBeTruthy();
  expect(queryByTestId('notes-button')).toBeTruthy();
});

// Sunday after Christmas Eve
test('On Sunday Dec 27, expect notes from Dec 27 (not video from Christmas Eve)', () => {
  const { queryByTestId } = render(
    <RecentTeaching teaching={mockTeaching[2]} note={mockNotes[2]} />
  );

  expect(queryByTestId('teaching-notes')).toBeTruthy();
  expect(queryByTestId('teaching-video')).toBeFalsy();
  expect(queryByTestId('notes-button')).toBeTruthy();
});

// One day after Christmas Eve
test('On Friday Dec 25, expect video from Christmas Eve, but no notes button', () => {
  const { queryByTestId } = render(
    <RecentTeaching teaching={mockTeaching[3]} note={mockNotes[3]} />
  );

  expect(queryByTestId('teaching-notes')).toBeFalsy();
  expect(queryByTestId('teaching-video')).toBeTruthy();
  expect(queryByTestId('notes-button')).toBeFalsy();
});
