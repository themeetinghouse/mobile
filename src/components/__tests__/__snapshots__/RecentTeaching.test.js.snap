// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recent Teaching component on Home Screen Snapshot using data from Sunday Dec 27 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "alignItems": "center",
        "backgroundColor": "#000000",
      },
    ]
  }
  testID="teaching-notes"
>
  <Image
    onError={[Function]}
    source={
      Object {
        "uri": "https://themeetinghouse.com/static/photos/series/adult-sunday-000.jpg",
      }
    }
    style={
      Object {
        "alignSelf": "center",
        "height": 192,
        "marginBottom": 24,
        "width": 159.975,
      }
    }
    testID="fallback-image"
  />
  <View
    style={
      Array [
        Object {},
        Object {
          "alignItems": "center",
          "marginHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "color": "#ffffff",
            "fontFamily": "Graphik-Bold-App",
            "fontSize": 32,
            "marginBottom": 8,
            "textAlign": "center",
          },
        ]
      }
      uppercase={false}
    >
      Tired Feet. Hopeful Hearts.
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "color": "#c8c8c8",
            "fontSize": 12,
            "marginBottom": 14,
          },
        ]
      }
      uppercase={false}
    >
      December 27, 2020
    </Text>
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      onPress={[Function]}
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "color": "#c8c8c8",
            "fontSize": 16,
            "lineHeight": 24,
            "marginBottom": 32,
            "textAlign": "center",
          },
        ]
      }
      testID="description"
      uppercase={false}
    >
      As Jesus people, we aren't the ones with all the answers - we're the ones with hope. We partner, wait, labour, and celebrate with the One who doesn't grow tired or weary as He ushers in his kingdom.
    </Text>
  </View>
  <View
    style={Object {}}
    testID="notes-button"
  >
    <View
      style={
        Array [
          Object {},
          Object {
            "paddingBottom": 10,
            "paddingTop": 10,
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "opacity": 1,
          }
        }
      >
        <Image
          source={1}
          square={true}
          style={
            Array [
              Object {
                "borderRadius": 0,
                "height": 56,
                "width": 56,
              },
              Object {
                "height": 24,
                "width": 24,
              },
              Object {
                "marginLeft": 0,
              },
            ]
          }
        />
        <View
          style={Object {}}
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "borderBottomColor": "#ffffff",
                  "borderBottomWidth": 1,
                  "color": "#ffffff",
                  "marginLeft": 15,
                  "marginRight": 15,
                  "paddingBottom": 3,
                  "paddingTop": 3,
                },
                Object {},
              ]
            }
            uppercase={false}
          >
            Notes
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Recent Teaching component on Home Screen Snapshot using null data 1`] = `
<View
  style={
    Array [
      Object {
        "padding": 15,
      },
      undefined,
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#ffffff"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
